### ISSUE TITLE
Autoclosable issue on multiple Metrics implementation

### ISSUE THREAD

**Issue by nickstolwijk** [micrometer-metrics/micrometer#1492]
Some Metrics extend AutoCloseable which causes code quality tools to flag those metrics as not being closed.

`
@Configuration
public class MetricsConfiguration {
    private final MeterRegistry registry;

    public MetricsConfiguration(MeterRegistry registry) {
        this.registry = registry;
    }

    @PostConstruct
    public void enableMetrics() {
        // do not change to try-with-resources as the resource will be removed on closing
        new JvmGcMetrics().bindTo(registry);
        new LogbackMetrics().bindTo(registry);
    }
}`

Is there a good reason to make these AutoCloseable? If not, could those be removed? Maybe replace it with a default method on MeterBinder for close?

**Comment by shakuzen** [comment]
Thank you for letting us know about the usability problem here. Indeed those `MeterBinder`s should not be closed after binding, but they generally should be closed on application shutdown, to clean up side effects of the binder. In environments that manage lifecycles (e.g. Spring beans), these being `AutoCloseable` can be helpful as the environment will automatically `close` them on application shutdown.

**Comment by nickstolwijk** [comment]
Would then Closeable be a better solution?

https://docs.oracle.com/javase/8/docs/api/java/io/Closeable.html

With regards,

Nick

**Comment by shakuzen** [comment]
I think `Closeable` would not match in these cases based on its JavaDoc:

> A Closeable is a source or destination of data that can be closed.

I'm happy to add to a JavaDoc on these classes or methods to help explain why and when they should be closed. What do you think?

**Comment by codemedian** [comment]
I've just run into the same issue and found it quite confusing. At the very least [this documentation page](https://micrometer.io/docs/ref/jvm) should contain some information about it and recommendations on how to close it.

**Comment by shakuzen** [comment]
@codemedian thank you for the suggestion. I've opened https://github.com/micrometer-metrics/micrometer-docs/issues/117

**Comment by codemedian** [comment]
I've just upgraded my micrometer version and I think the way this is done is a bit odd/inconsistent. 

I understand that you want to close the resources and all, but there's no clean way to hook into micrometer. I.e. the registry doesn't have a shutdown hook and/or cannot be shutdown. This effectively means that in most cases people will do something along those lines

```java
JvmGcMetrics jvmGcMetrics = new JvmGcMetrics();
jvmGcMetrics.bindTo(prometheusRegistry);

Runtime.getRuntime().addShutdownHook(new Thread(jvmGcMetrics::close));```

So I think that a better way would be to either have that JvmGcMetrics thing clean up itself by adding above shutdown hook, or providing a clean way to tear down a registry (which afaik isn't possible atm)

**Comment by christian-esken** [comment]
I agree with @codemedian . As long as JvmGcMetrics is AutoCloseable it can be cleaned up by a shutdown hook. And Sonar can then be quietened (for me it is "java:S2095"), including a comment explaining why the warning is suppressed.

        @SuppressWarnings("java:S2095")
        // Do not change JvmGcMetrics to try-with-resources as the JvmGcMetrics will not be available after (auto-)closing.
        // See https://github.com/micrometer-metrics/micrometer/issues/1492
        JvmGcMetrics jvmGcMetrics = new JvmGcMetrics();

**Comment by marcingrzejszczak** [comment]
So would adding additional documentation about adding `Runtime.getRuntime().addShutdownHook(new Thread(jvmGcMetrics::close));`  hook would be sufficient to close this?

**Comment by christian-esken** [comment]
While I am not fully aware of the impacts of a change (e.g. from AutoCloseable to Closeable), I can imagine that it impacts runtime behavior too much, as seen in https://github.com/micrometer-metrics/micrometer/pull/1760 for SpringBoot.

Thus, documenting would be OK for me.
